### PR TITLE
Issue 4885: Add component sdk

### LINKED ISSUE #4885: Investigate a component SDK for dmSDK
It would be beneficial to have an sdk for adding/removing/replacing components.
It would allow us to move core functionality out to extensions, thus decoupling the code a bit.
Also, it's a potential way to remove more payload in the engine, by removing components that aren't used in the game.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/gameobject/src/gameobject/component.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+#include <dlib/static_assert.h>
+
+namespace dmGameObject
+{
+
+static ComponentTypeDescriptor g_ComponentTypeSentinel = {0};
+
+Result RegisterComponentTypeDescriptor(ComponentTypeDescriptor* desc, const char* name, ComponentTypeCreateFunction create_fn)
+{
```
Comment (jhonnyking) [inline engine/gameobject/src/gameobject/component.cpp:23]: There's no namespace indentation here

File: engine/gameobject/src/gameobject/gameobject_comp.cpp [hunk 2]
```diff
@@ -1,57 +1,64 @@
...
-        anim_component.m_Name = "animc";
-        anim_component.m_Context = 0x0;
-        anim_component.m_NewWorldFunction = &CompAnimNewWorld;
-        anim_component.m_DeleteWorldFunction = &CompAnimDeleteWorld;
-        anim_component.m_AddToUpdateFunction = &CompAnimAddToUpdate;
-        anim_component.m_ReadsTransforms = 1;
-        anim_component.m_UpdateFunction = &CompAnimUpdate;
-        anim_component.m_UpdateOrderPrio = 250;
-        return RegisterComponentType(regist, anim_component);
+        dmLogWarning("MAWE: component '%s' created", type->m_Name);
```
Comment (jhonnyking) [inline engine/gameobject/src/gameobject/gameobject_comp.cpp:23]: Should this be here?

**Review by jhonnyking — DISMISSED** [review]
It's a huuge change and I can't really go through it all that detailed. But in any case I think it looks good.

File: engine/gameobject/src/gameobject/gameobject_comp.cpp (hunk 2 above)
Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject_comp.cpp:23]: heh, no it shouldn't :)

**Pull request merged**